enriched-span-constants/src/main/java/org/hypertrace/traceenricher/enrichedspan/constants/utils/EnrichedSpanUtils.java

```diff
@@ -301,12 +317,74 @@ public static String getUserAgent(Event event) {
     return Optional.empty();
   }
 
+  /**
+   * This method returns an Optional which, if present, contains the full URL, including the protocol and host name.
+   * If the tags in the raw span do not contain the full URL, it tries to construct a full URL using the protocol
+   * ("http") and host name.
+   */
   public static Optional<String> getFullHttpUrl(Event event) {
-    if (event.getHttp() != null && event.getHttp().getRequest() != null) {
-      return Optional.ofNullable(event.getHttp().getRequest().getUrl());
+    Optional<String> optionalUrl = Optional.ofNullable(event.getHttp())
+            .map(org.hypertrace.core.datamodel.eventfields.http.Http::getRequest)
+            .map(Request::getUrl);
+
+    if (optionalUrl.isEmpty() && event.getAttributes() != null && event.getAttributes().getAttributeMap() != null) {
+      // iterate on all URL related attribute keys and find / create one with protocol, host header and path
+      Map<String, AttributeValue> attributesMap = event.getAttributes().getAttributeMap();
+      for (String urlAttribute : FULL_URL_ATTRIBUTES) {
+        if (attributesMap.containsKey(urlAttribute)) {
+          String url = attributesMap.get(urlAttribute).getValue();
+          if (url != null && url.toLowerCase().startsWith(HTTP_PREFIX)) {
+            optionalUrl = Optional.of(url);
+          } else {
+            optionalUrl = constructHttpUrlFromHostName(event);
+          }
+        }
+      }
     }
+    return optionalUrl;
+  }
+
+  private static Optional<String> constructHttpUrlFromHostName(Event event) {
+    Optional<String> optionalUrl = Optional.empty();
+    Optional<RequestHeaders> requestHeaders = Optional.ofNullable(event.getHttp())
+            .map(org.hypertrace.core.datamodel.eventfields.http.Http::getRequest)
+            .map(Request::getHeaders);
+    Optional<String> hostName = requestHeaders.map(RequestHeaders::getHost);
+    Optional<String> path = requestHeaders.map(RequestHeaders::getPath);
+    if (hostName.isPresent() && path.isPresent()) {
+      try {
+        URL fullUrl = new URL(HTTP_PREFIX, hostName.get(), path.get());
+        optionalUrl = Optional.of(fullUrl.toString());
+      } catch (MalformedURLException e) {
+        // ignore. optionalUrl will be empty.
+      }
+    }
+    return optionalUrl;
+  }
 
-    return Optional.empty();
+  /**
+   * This method returns an Optional which, if present, contains the URL as captured in the tags. It may or may not
+   * be a full URL. For instance, it may only contain the path and not the protocol and hostname like
+   * "/customer?customer=392".
+   */
+  public static Optional<String> getHttpUrl(Event event) {
```

Comment:
@GurtejSohi I think the behavior of `event.getHttp().getRequest().getUrl()` should be same as this method. if not, I think it's better to rather fix that method than adding this new method so that everyone always can access non-null HTTP url, full URL or not from the span.

Reply:
+1